pool.go

```diff
@@ -13,17 +15,66 @@ func newBufPool(depth, bufLen int) *bufPool {
 }
 
 func (p *bufPool) Get() []byte {
-	select {
-	case b := <-p.ch:
-		return b
-	default:
+	if p == nil {
+		// functional default: no reuse.
 		return make([]byte, p.blen)
 	}
+
+	for {
+		select {
+		case b := <-p.ch:
+			if cap(b) < p.blen {
+				// just in case: throw away any buffer with insufficient capacity.
+				continue
+			}
+
+			return b[:p.blen]
+
+		default:
+			return make([]byte, p.blen)
+		}
+	}
 }
 
 func (p *bufPool) Put(b []byte) {
+	if p == nil {
+		// functional default: no reuse.
+		return
+	}
+
+	if cap(b) < p.blen || cap(b) > p.blen*2 {
+		// DO NOT reuse buffers with insufficient capacity.
+		// This could cause panics when resizing to p.blen.
+
+		// DO NOT reuse buffers with excessive capacity.
+		// This could cause memory leaks.
+		return
+	}
+
 	select {
 	case p.ch <- b:
 	default:
 	}
 }
+
+type resChanPool chan chan result
```

Comment:
This is a fun channel of channels to keep a convenient store of `result` channels. This saves more on `allocs` than it does in memory, but it is really big to not have so much churn of these channels. If they were able to be unbuffered, I would see reason to pool them for reuse, as then they wouldn’t have backing memory. But these keep getting allocated to the stack and causes lots of churn.